File: src/DisplayPanel.java
import blocks.Block;
import blocks.IBlock;

import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

class DisplayPanel extends JPanel {
    ArrayList<Level> levelArray = GameLogic.levelArray;
    int currentLevel = GameLogic.currentLevel;
    BufferedImage[] BackgroundImages = GameLogic.BackgroundImages;
    BufferedImage[] blockImages = GameLogic.blockImages;
    BufferedImage[] tileImages = GameLogic.tileImages;
    Mob character =  GameLogic.character;

    public void paintComponent(Graphics g) {

        super.paintComponent(g);

        switch (GameLogic.current_gamestate) {
            case main_screen:
                drawMainScreen(g);
                break;
            case gameplay_screen:
                drawGameScreen(g);
                break;
        }
    }

    public void drawGameScreen(Graphics g) {
        drawingBackground(g);
        drawingTiles(g);
        drawingBlocks(g);
        drawingCharacter(g);
        drawingMobs(g);
        drawingGrapple(g);
        drawingAimingCircle(g);
        //g.drawImage(testanimation[cycle%testanimation.length], 40, 40, this);
    }

    private void drawingAimingCircle(Graphics g) {
        int charCenterX = character.getX() + (character.getWidth() / 2);
        int charCenterY = character.getY() + (character.getHeight() / 2);

        int r = -28;

        int vectorx = character.getCircleProjection(charCenterX, charCenterY,GameLogic. mouseX, GameLogic.mouseY, r);
        int vectory = character.getCircleProjection(charCenterY, charCenterX,GameLogic. mouseY, GameLogic.mouseX, r);

        //create a vector by b-a for points
        g.setColor(Color.RED);
        g.fillOval(vectorx - 5, vectory - 5, 10, 10);
    }

    private void drawingBackground(Graphics g) {
        if (levelArray.get(currentLevel).getBackground().getImage().equals("graphics/sword-and-sworcery.png")) {
            g.drawImage(BackgroundImages[0],
                    levelArray.get(currentLevel).getBackground().getX(),
                    levelArray.get(currentLevel).getBackground().getY(),
                    null);
        }
        if (levelArray.get(currentLevel).getBackground().getImage().equals("graphics/windows_xp_bliss-wide.jpg"))
            g.drawImage(BackgroundImages[1],
                    levelArray.get(currentLevel).getBackground().getX(),
                    levelArray.get(currentLevel).getBackground().getY(),
                    null);
        if (levelArray.get(currentLevel).getBackground().getImage().equals("graphics/whiteBackground.png"))
            g.drawImage(BackgroundImages[2],
                    levelArray.get(currentLevel).getBackground().getX(),
                    levelArray.get(currentLevel).getBackground().getY(),
                    null);
    }

    private void drawingMobs(Graphics g) {
        for (Mob m : levelArray.get(currentLevel).getLevelMobs())
            g.drawImage(m.getSprite(), m.getX(), m.getY(), this);
    }

    private void drawingCharacter(Graphics g) {
        if (character.getJumpable() == 0) {
            if (character.getXspeed() < 0) character.setState(1);
            if (character.getXspeed() == 0) character.setState(0);
            if (character.getXspeed() < 0) character.setState(2);
        } else character.setState(3);

        g.drawImage(character.getSprite(), character.getX(), character.getY(), this);
    }
    private void drawingBlocks(Graphics g) {
        for (IBlock b : levelArray.get(currentLevel).getLevelBlocks()) {
            if (b.getType() == Block.Type.PLAIN)
                g.drawImage(blockImages[0], b.getX(), b.getY(), this);
            if (b.getType()== Block.Type.DEATH)
                g.drawImage(blockImages[1], b.getX(), b.getY(), this);
            if (b.getType()== Block.Type.SLOW)
                g.drawImage(blockImages[2], b.getX(), b.getY(), this);
            if (b.getType()== Block.Type.PLATFORM)
                g.drawImage(blockImages[3], b.getX(), b.getY(), this);
            if (b.getType()== Block.Type.PORTAL)
                g.drawImage(blockImages[4], b.getX(), b.getY(), this);
        }
    }

    private void drawingTiles(Graphics g) {
        for (Tile t : levelArray.get(currentLevel).getLevelTiles()) {
            if (t.getImage().equals("door"))
                g.drawImage(tileImages[0], t.getX(), t.getY(), this);
            if (t.getImage().equals("grass"))
                g.drawImage(tileImages[1], t.getX(), t.getY(), this);
            if (t.getImage().equals("topright"))
                g.drawImage(tileImages[2], t.getX(), t.getY(), this);
            if (t.getImage().equals("topleft"))
                g.drawImage(tileImages[3], t.getX(), t.getY(), this);
            if (t.getImage().equals("horizontal"))
                g.drawImage(tileImages[4], t.getX(), t.getY(), this);
            if (t.getImage().equals("bottomright"))
                g.drawImage(tileImages[5], t.getX(), t.getY(), this);
            if (t.getImage().equals("bottomleft"))
                g.drawImage(tileImages[6], t.getX(), t.getY(), this);
            if (t.getImage().equals("bottomhorizontal"))
                g.drawImage(tileImages[7], t.getX(), t.getY(), this);
        }
    }

    public void drawMainScreen(Graphics g) {
        g.drawImage(BackgroundImages[0],
                levelArray.get(currentLevel).getBackground().getX(),
                levelArray.get(currentLevel).getBackground().getY(),
                null);
    }

    private void drawingGrapple(Graphics g) {
        if (character.getTethered()) {
            g.setColor(Color.black);
            g.drawLine(character.getX() + character.getWidth() / 2,
                    character.getY() + character.getHeight() / 2,
                    character.getWeaponArray().get(0).getHitX(),
                    character.getWeaponArray().get(0).getHitY());
        }
    }
}

File: src/DisplayPanel.kt
import GameLogic.Gamestate.GAMEPLAY_SCREEN
import GameLogic.Gamestate.MAIN_SCREEN
import GameLogic.Gamestate.TITLE_SCREEN
import common.Block
import common.IBlock
import common.ITile
import java.awt.Color
import java.awt.Graphics
import javax.swing.JPanel

class DisplayPanel : JPanel() {
    public override fun paintComponent(g: Graphics) {
        super.paintComponent(g)

        when (GameLogic.current_gamestate) {
            MAIN_SCREEN -> drawMainScreen(g)
            GAMEPLAY_SCREEN -> drawGameScreen(g)
            TITLE_SCREEN -> TODO()
        }
    }

    private fun drawGameScreen(g: Graphics) {
        drawingBackground(g)
        drawingTiles(g)
        drawingCharacter(g)
        drawingMobs(g)
        drawingGrapple(g)
        drawingAimingCircle(g)
        //gameLogic.drawImage(testanimation[cycle%testanimation.length], 40, 40, this);
    }

    private fun drawingAimingCircle(g: Graphics) {
        val charCenterX = GameLogic.character.x + GameLogic.character.width / 2
        val charCenterY = GameLogic.character.y + GameLogic.character.height / 2
        val r = -28
        val vectorx = GameLogic.character.getCircleProjection(charCenterX, charCenterY, GameLogic.mouseX, GameLogic.mouseY, r)
        val vectory = GameLogic.character.getCircleProjection(charCenterY, charCenterX, GameLogic.mouseY, GameLogic.mouseX, r)
        //create a vector by b-a for points
        g.color = Color.RED
        g.fillOval(vectorx - 5, vectory - 5, 10, 10)
    }

    private fun drawingBackground(g: Graphics) {
        val background = GameLogic.levelArray[GameLogic.currentLevel].background
        when (background.image) {
            "sword-and-sworcery.png" -> g.drawImage(backgroundImages[0], background.x, background.y, null)
            "windows_xp_bliss-wide.jpg" -> g.drawImage(backgroundImages[1], background.x, background.y, null)
            "whiteBackground.png" -> g.drawImage(backgroundImages[2], background.x, background.y, null)
        }
    }

    private fun drawingMobs(g: Graphics) {
        GameLogic.levelArray[GameLogic.currentLevel].levelMobs.forEach {
            g.drawImage(it.sprite, it.x, it.y, this)
        }
    }

    private fun drawingCharacter(g: Graphics) {
        if (GameLogic.character.jumpable == 0) {
            when {
                GameLogic.character.xspeed < 0 -> GameLogic.character.state = 1
                GameLogic.character.xspeed == 0 -> GameLogic.character.state = 0
                GameLogic.character.xspeed < 0 -> GameLogic.character.state = 2
            }
        } else {
            GameLogic.character.state = 3
        }

        g.drawImage(GameLogic.character.sprite, GameLogic.character.x, GameLogic.character.y, this)
    }

    private fun drawingTiles(g: Graphics) {
        fun draw(tile: ITile, index: String) {
            g.drawImage(images[index]?.value, tile.x, tile.y, this)
        }
        GameLogic.levelArray[GameLogic.currentLevel].levelTiles.forEach {
            if (it is IBlock) {
                draw(it, it.type.toString())
            } else {
                draw(it, it.image!!)
            }
        }
    }

    private fun drawMainScreen(g: Graphics) {
        val level = GameLogic.levelArray[GameLogic.currentLevel]
        g.drawImage(
                backgroundImages[0],
                level.background.x,
                level.background.y,
                null
        )
    }

    private fun drawingGrapple(g: Graphics) {
        if (GameLogic.character.tethered) {
            g.color = Color.black
            g.drawLine(
                    GameLogic.character.x + GameLogic.character.width / 2,
                    GameLogic.character.y + GameLogic.character.height / 2,
                    GameLogic.character.weaponArray[0].hitX,
                    GameLogic.character.weaponArray[0].hitY
            )
        }
    }

    companion object {
        var images = mapOf(
                Block.Type.PLAIN.toString() to lazy { ResourceLoader.getImage(Block(0, 0, 0, 0, Block.Type.PLAIN).image!!) },
                Block.Type.DEATH.toString() to lazy { ResourceLoader.getImage(Block(0, 0, 0, 0, Block.Type.DEATH).image!!) },
                Block.Type.SLOW.toString() to lazy { ResourceLoader.getImage(Block(0, 0, 0, 0, Block.Type.SLOW).image!!) },
                Block.Type.PLATFORM.toString() to lazy { ResourceLoader.getImage(Block(0, 0, 0, 0, Block.Type.PLATFORM).image!!) },
                Block.Type.PORTAL.toString() to lazy { ResourceLoader.getImage(Block(0, 0, 0, 0, Block.Type.PORTAL).image!!) },
                "door" to lazy { ResourceLoader.getImage("Door01.png") },
                "grass" to lazy { ResourceLoader.getImage("GrassBlock01.png") },
                "topright" to lazy { ResourceLoader.getImage("CornerBlockTopRight.png") },
                "topleft" to lazy { ResourceLoader.getImage("CornerBlockTopLeft.png") },
                "horizontal" to lazy { ResourceLoader.getImage("Horizontal.png") },
                "bottomright" to lazy { ResourceLoader.getImage("CornerBlockBottomRight.png") },
                "bottomleft" to lazy { ResourceLoader.getImage("CornerBlockBottomLeft.png") },
                "bottomhorizontal" to lazy { ResourceLoader.getImage("BottomHorizontal.png") }
        )
        var backgroundImages = arrayOf(
                ResourceLoader.getImage("sword-and-sworcery.png"),
                ResourceLoader.getImage("windows_xp_bliss-wide.jpg"),
                ResourceLoader.getImage("whiteBackground.png")
        )
    }
}

